Adding Hibernate panache dependency + driver and no config leads to error

When I add Hibernate Panache as a dependency, and do `mvn compile quarkus:dev`, I get 

```
        [error]: Build step io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor#doParseAndRegisterSubstrateResources threw an exception: io.quark
us.deployment.configuration.ConfigurationError: Hibernate extension cannot guess the dialect as no JDBC driver is specified by 'quarkus.datasource.driv
er'
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:134)
        at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:105)
        at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: org.jboss.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor#doParseAndRegisterSubstrateResources threw an exception: io.quark
us.deployment.configuration.ConfigurationError: Hibernate extension cannot guess the dialect as no JDBC driver is specified by 'quarkus.datasource.driv
er'
        at org.jboss.builder.Execution.run(Execution.java:123)
        at org.jboss.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:136)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:110)
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:99)
        ... 2 more
Caused by: io.quarkus.deployment.configuration.ConfigurationError: Hibernate extension cannot guess the dialect as no JDBC driver is specified by 'quar
kus.datasource.driver'
        at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.guessDialect(HibernateOrmProcessor.java:473)
        at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.handleHibernateORMWithNoPersistenceXml(HibernateOrmProcessor.java:330)
        at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.doParseAndRegisterSubstrateResources(HibernateOrmProcessor.java:127)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:506)
        at org.jboss.builder.BuildContext.run(BuildContext.java:413)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1998)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1525)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1416)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

Note that I have no `@Entity` annotations. I think we did fix that issue in the past @gsmet but it might have resurfaced.

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
@@ -23,6 +23,7 @@ import java.io.IOException;
 import java.net.URL;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
@@ -116,38 +117,27 @@ public final class HibernateOrmProcessor {
     }
 
     @BuildStep
-    void doParseAndRegisterSubstrateResources(BuildProducer<PersistenceUnitDescriptorBuildItem> persistenceProducer,
-            BuildProducer<SubstrateResourceBuildItem> resourceProducer,
-            BuildProducer<HotDeploymentConfigFileBuildItem> hotDeploymentProducer,
-            BuildProducer<SystemPropertyBuildItem> systemPropertyProducer,
-            ArchiveRootBuildItem root,
-            ApplicationArchivesBuildItem applicationArchivesBuildItem,
-            Optional<DataSourceDriverBuildItem> driverBuildItem) throws IOException {
-        List<ParsedPersistenceXmlDescriptor> descriptors = loadOriginalXMLParsedDescriptors();
-        handleHibernateORMWithNoPersistenceXml(descriptors, resourceProducer, hotDeploymentProducer, systemPropertyProducer,
-                root, driverBuildItem, applicationArchivesBuildItem);
-        for (ParsedPersistenceXmlDescriptor i : descriptors) {
-            persistenceProducer.produce(new PersistenceUnitDescriptorBuildItem(i));
-        }
-    }
-
-    @BuildStep
     @Record(STATIC_INIT)
     public void build(RecorderContext recorder, HibernateOrmTemplate template,
-            List<PersistenceUnitDescriptorBuildItem> descItems,
             List<AdditionalJpaModelBuildItem> additionalJpaModelBuildItems,
             List<NonJpaModelBuildItem> nonJpaModelBuildItems,
             CombinedIndexBuildItem index,
             ApplicationIndexBuildItem applicationIndex,
-            BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
+            ArchiveRootBuildItem archiveRoot,
+            ApplicationArchivesBuildItem applicationArchivesBuildItem,
+            Optional<DataSourceDriverBuildItem> driverBuildItem,
             BuildProducer<FeatureBuildItem> feature,
+            BuildProducer<PersistenceUnitDescriptorBuildItem> persistenceUnitDescriptorProducer,
+            BuildProducer<SubstrateResourceBuildItem> resourceProducer,
+            BuildProducer<HotDeploymentConfigFileBuildItem> hotDeploymentProducer,
+            BuildProducer<SystemPropertyBuildItem> systemPropertyProducer,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
             BuildProducer<JpaEntitiesBuildItem> domainObjectsProducer,
             BuildProducer<BeanContainerListenerBuildItem> beanContainerListener) throws Exception {
 
         feature.produce(new FeatureBuildItem(FeatureBuildItem.HIBERNATE_ORM));
 
-        List<ParsedPersistenceXmlDescriptor> descriptors = descItems.stream()
-                .map(PersistenceUnitDescriptorBuildItem::getDescriptor).collect(Collectors.toList());
+        List<ParsedPersistenceXmlDescriptor> explicitDescriptors = loadOriginalXMLParsedDescriptors();
 
         // build a composite index with additional jpa model classes
         Indexer indexer = new Indexer();
@@ -161,7 +151,8 @@ public final class HibernateOrmProcessor {
         Set<String> nonJpaModelClasses = nonJpaModelBuildItems.stream()
                 .map(NonJpaModelBuildItem::getClassName)
                 .collect(Collectors.toSet());
-        JpaJandexScavenger scavenger = new JpaJandexScavenger(reflectiveClass, descriptors, compositeIndex, nonJpaModelClasses);
+        JpaJandexScavenger scavenger = new JpaJandexScavenger(reflectiveClass, explicitDescriptors, compositeIndex,
+                nonJpaModelClasses);
         final JpaEntitiesBuildItem domainObjects = scavenger.discoverModelAndRegisterForReflection();
 
         // remember how to run the enhancers later
@@ -172,17 +163,28 @@ public final class HibernateOrmProcessor {
             return;
         }
 
-        for (String className : domainObjects.getClassNames()) {
+        template.callHibernateFeatureInit();
+
+        // handle the implicit persistence unit
+        List<ParsedPersistenceXmlDescriptor> allDescriptors = new ArrayList<>(explicitDescriptors.size() + 1);
+        allDescriptors.addAll(explicitDescriptors);
+        handleHibernateORMWithNoPersistenceXml(allDescriptors, resourceProducer, hotDeploymentProducer, systemPropertyProducer,
+                archiveRoot, driverBuildItem, applicationArchivesBuildItem);
+
+        for (ParsedPersistenceXmlDescriptor descriptor : allDescriptors) {
+            persistenceUnitDescriptorProducer.produce(new PersistenceUnitDescriptorBuildItem(descriptor));
+        }
+
+        for (String className : domainObjects.getEntityClassNames()) {
             template.addEntity(className);
         }
         template.enlistPersistenceUnit();
-        template.callHibernateFeatureInit();
 
         //set up the scanner, as this scanning has already been done we need to just tell it about the classes we
         //have discovered. This scanner is bytecode serializable and is passed directly into the template
         QuarkusScanner scanner = new QuarkusScanner();
         Set<ClassDescriptor> classDescriptors = new HashSet<>();
-        for (String i : domainObjects.getClassNames()) {
+        for (String i : domainObjects.getAllModelClassNames()) {
             QuarkusScanner.ClassDescriptorImpl desc = new QuarkusScanner.ClassDescriptorImpl(i,
                     ClassDescriptor.Categorization.MODEL);
             classDescriptors.add(desc);
@@ -192,7 +194,7 @@ public final class HibernateOrmProcessor {
         //now we serialize the XML and class list to bytecode, to remove the need to re-parse the XML on JVM startup
         recorder.registerNonDefaultConstructor(ParsedPersistenceXmlDescriptor.class.getDeclaredConstructor(URL.class),
                 (i) -> Collections.singletonList(i.getPersistenceUnitRootUrl()));
-        beanContainerListener.produce(new BeanContainerListenerBuildItem(template.initMetadata(descriptors, scanner)));
+        beanContainerListener.produce(new BeanContainerListenerBuildItem(template.initMetadata(allDescriptors, scanner)));
     }
 
     @BuildStep
@@ -297,7 +299,7 @@ public final class HibernateOrmProcessor {
     }
 
     private boolean hasEntities(JpaEntitiesBuildItem jpaEntities, List<NonJpaModelBuildItem> nonJpaModels) {
-        return !jpaEntities.getClassNames().isEmpty() || !nonJpaModels.isEmpty();
+        return !jpaEntities.getEntityClassNames().isEmpty() || !nonJpaModels.isEmpty();
     }
 
     private boolean isUserDefinedProducerMissing(IndexView index, DotName annotationName) {
@@ -483,7 +485,7 @@ public final class HibernateOrmProcessor {
             List<AdditionalJpaModelBuildItem> additionalJpaModelBuildItems,
             BuildProducer<GeneratedClassBuildItem> additionalClasses) {
         HibernateEntityEnhancer hibernateEntityEnhancer = new HibernateEntityEnhancer();
-        for (String i : domainObjects.getClassNames()) {
+        for (String i : domainObjects.getAllModelClassNames()) {
             transformers.produce(new BytecodeTransformerBuildItem(i, hibernateEntityEnhancer));
         }
         for (AdditionalJpaModelBuildItem additionalJpaModel : additionalJpaModelBuildItems) {

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaEntitiesBuildItem.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaEntitiesBuildItem.java
@@ -19,6 +19,8 @@ package io.quarkus.hibernate.orm.deployment;
 import java.util.HashSet;
 import java.util.Set;
 
+import javax.persistence.Entity;
+
 import org.jboss.builder.item.SimpleBuildItem;
 
 import io.quarkus.deployment.annotations.BuildProducer;
@@ -30,19 +32,35 @@ import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
  */
 public final class JpaEntitiesBuildItem extends SimpleBuildItem {
 
-    private final Set<String> classNames = new HashSet<String>();
+    private final Set<String> entityClassNames = new HashSet<String>();
+    private final Set<String> allModelClassNames = new HashSet<String>();
 
-    void addEntity(final String className) {
-        classNames.add(className);
+    void addEntityClass(final String className) {
+        entityClassNames.add(className);
+        allModelClassNames.add(className);
+    }
+
+    void addModelClass(final String className) {
+        allModelClassNames.add(className);
     }
 
     void registerAllForReflection(final BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
-        for (String className : classNames) {
+        for (String className : allModelClassNames) {
             reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, className));
         }
     }
 
-    public Set<String> getClassNames() {
-        return classNames;
+    /**
+     * @return the list of entities (i.e. classes marked with {@link Entity})
+     */
+    public Set<String> getEntityClassNames() {
+        return entityClassNames;
+    }
+
+    /**
+     * @return the list of all model class names: entities, mapped super classes...
+     */
+    public Set<String> getAllModelClassNames() {
+        return allModelClassNames;
     }
 }

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java
@@ -29,7 +29,6 @@ import javax.persistence.Embedded;
 import javax.persistence.Entity;
 import javax.persistence.MappedSuperclass;
 
-import org.graalvm.nativeimage.ImageInfo;
 import org.hibernate.jpa.boot.internal.ParsedPersistenceXmlDescriptor;
 import org.hibernate.jpa.boot.spi.PersistenceUnitDescriptor;
 import org.jboss.jandex.AnnotationInstance;
@@ -63,17 +62,17 @@ final class JpaJandexScavenger {
 
     private static final DotName ENUM = DotName.createSimple(Enum.class.getName());
 
-    private final List<ParsedPersistenceXmlDescriptor> descriptors;
+    private final List<ParsedPersistenceXmlDescriptor> explicitDescriptors;
     private final BuildProducer<ReflectiveClassBuildItem> reflectiveClass;
     private final IndexView indexView;
     private final Set<String> nonJpaModelClasses;
 
     JpaJandexScavenger(BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
-            List<ParsedPersistenceXmlDescriptor> descriptors,
+            List<ParsedPersistenceXmlDescriptor> explicitDescriptors,
             IndexView indexView,
             Set<String> nonJpaModelClasses) {
         this.reflectiveClass = reflectiveClass;
-        this.descriptors = descriptors;
+        this.explicitDescriptors = explicitDescriptors;
         this.indexView = indexView;
         this.nonJpaModelClasses = nonJpaModelClasses;
     }
@@ -90,7 +89,7 @@ final class JpaJandexScavenger {
         enlistJPAModelClasses(indexView, domainObjectCollector, enumTypeCollector, MAPPED_SUPERCLASS, unindexedClasses);
         enlistReturnType(indexView, domainObjectCollector, enumTypeCollector, unindexedClasses);
 
-        for (PersistenceUnitDescriptor pud : descriptors) {
+        for (PersistenceUnitDescriptor pud : explicitDescriptors) {
             enlistExplicitClasses(indexView, domainObjectCollector, enumTypeCollector, pud.getManagedClassNames(),
                     unindexedClasses);
         }
@@ -122,12 +121,11 @@ final class JpaJandexScavenger {
         for (String className : managedClassNames) {
             DotName dotName = DotName.createSimple(className);
             boolean isInIndex = index.getClassByName(dotName) != null;
-            if (isInIndex) {
-                addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, dotName, unindexedClasses);
-            } else {
+            if (!isInIndex) {
                 unindexedClasses.add(dotName);
-                domainObjectCollector.addEntity(className);
             }
+
+            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, dotName, unindexedClasses);
         }
     }
 
@@ -175,7 +173,7 @@ final class JpaJandexScavenger {
             }
             addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, targetDotName,
                     unindexedClasses);
-            domainObjectCollector.addEntity(targetDotName.toString());
+            collectDomainObject(domainObjectCollector, klass);
         }
     }
 
@@ -208,7 +206,8 @@ final class JpaJandexScavenger {
         }
 
         //Capture this one (for various needs: Reflective access enablement, Hibernate enhancement, JPA Template)
-        domainObjectCollector.addEntity(className.toString());
+        collectDomainObject(domainObjectCollector, classInfo);
+
         // add superclass recursively
         addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, classInfo.superName(),
                 unindexedClasses);
@@ -216,6 +215,14 @@ final class JpaJandexScavenger {
         for (DotName interfaceDotName : classInfo.interfaceNames()) {
             addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, interfaceDotName,
                     unindexedClasses);
+        }
+    }
+
+    private static void collectDomainObject(JpaEntitiesBuildItem domainObjectCollector, ClassInfo modelClass) {
+        if (modelClass.classAnnotation(JPA_ENTITY) != null) {
+            domainObjectCollector.addEntityClass(modelClass.name().toString());
+        } else {
+            domainObjectCollector.addModelClass(modelClass.name().toString());
         }
     }
 }

--- a/extensions/panache/hibernate-orm-panache/deployment/pom.xml
+++ b/extensions/panache/hibernate-orm-panache/deployment/pom.xml
@@ -54,15 +54,10 @@
             <artifactId>asm-analysis</artifactId>
         </dependency>
 
+        <!-- test dependencies -->
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>io.quarkus.gizmo</groupId>
-            <artifactId>gizmo</artifactId>
-            <type>test-jar</type>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5-internal</artifactId>
             <scope>test</scope>
         </dependency>
     </dependencies>

--- a/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/test/NoConfigTest.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/test/NoConfigTest.java
@@ -1,0 +1,20 @@
+package io.quarkus.hibernate.orm.panache.test;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class NoConfigTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest().setArchiveProducer(
+            () -> ShrinkWrap.create(JavaArchive.class));
+
+    @Test
+    public void testNoConfig() {
+        // we should be able to start the application, even with no configuration at all
+    }
+}

